src/doc/rustc/src/platform-support/bpf-unknown-none.md

```diff
@@ -0,0 +1,123 @@
+# `bpf*-unknown-none`
+
+**Tier: 3**
+
+* `bpfeb-unknown-none` (big endian)
+* `bpfel-unknown-none` (little endian)
+
+Targets for the [eBPF virtual machine][ebpf].
+
+## Target maintainers
+
+- [@dave-tucker](https://github.com/dave-tucker)
+- [@vadorovsky](https://github.com/vadorovsky)
+
+## Requirements
+
+BPF targets require a Rust toolchain with `rust-src` component and
+[bpf-linker][bpf-linker].
+
+They don't support std and alloc and are meant for `no_std` environment.
+
+`extern "C"` uses the [BPF ABI calling convention][bpf-abi].
+
+Produced binaries use the ELF format.
+
+## Building the target
+
+You can build Rust with support for BPF targets by adding them to the `target`
+list in `config.toml`:
+
+```toml
+[build]
+target = ["bpfeb-unknown-none", "bpfel-unknown-none"]
+```
+
+## Building Rust programs
+
+Rust does not yet ship pre-compiled artifacts for this target. To compile for
+this target, you will either need to build Rust with the target enabled (see
+"Building the target" above), or build your own copy of `core` by using
+`build-std` or similar.
+
+Building the BPF target requires specifying it explicitly. Users can either
+add them to the `target` list in `config.toml`:
+
+```toml
+[build]
+target = ["bpfel-unknown-none"]
+```
+
+Or specify it directly in the `cargo build` invocation:
+
+```console
+cargo +nightly build -Z build-std=core --target bpfel-unknown-none
+```
+
+## Testing
+
+eBPF bytecode need to be executed on an eBPF virtual machine, like the one
+provided by the Linux kernel or one of user-space implementations like
+[rbpf][rbpf]. None of them supports running Rust testsuite.
+
+Currently the best way of unit testing the code used in BPF programs is
+extracting it to a separate crate, which can be built and ran on a host target,
+and using it as a dependency.
+
+## Cross-compilation toolchains
+
+Technically speaking, every build for BPF target is a cross build and is
+performed on a host target (e.g. `x86_64-unknown-linux-*`) for the BPF target
+(e.g. `bpfel-unknown-none`). The resulting BPF bytecode can be executed on a
+virtual machine, which can run on different host architectures (x86_64, aarch64,
+riscv64 etc.). The BPF bytecode format stays the same regardless of the host
+architecture hosting the virtual machine. When building BPF code, compiler does
+not care about the architecture hosting the VM.
```

Comment:
Why is there both LE eBPF and BE eBPF, then?

Does the BPF VM host not have to match the code in endianness, or is it just indicative of whether the VM should perform appropriate byte-swapping?

Reply:
I just double-checked and the former is right - BPF VM host's endianness needs to match. So the bytecode is portable only across architectures of the same endianness.

Reply:
Personally, I wouldn't say anything about portability, I'd just say that the bytecode is loaded into the kernel and executed in a vm (often getting JITted first).

Bytecode portability seems like an irrelevant detail, and only really matters for programs that don't interact with the kernel (eg XDP). As soon as you interface with the kernel, nothing is portable anymore, not even across same host endianness.

Reply:
What I mostly wanted to highlight here is that the host architecture differences are not the direct concern of the compiler.

Reply:
Yes but my worry is that if someone starts tinkering with eBPF from rust and start from this doc, without knowing anything about aya, they might get the wrong idea that the bytecode is portable, when in fact for the majority of eBPF program types it isn't.

Reply:
Fair, I won't mention portability and rephrase the whole thing to "Programs are always cross-compiled from a host to the eBPF target. The architecture of the eBPF VM host matters for programs (because of bpf helpers, kfuncs, pt_regs etc.), but it's not the concern of the compiler. Instead, it's the developer's responsibility to take care of compiling programs with appropriate types per architecture. Aya handles that by providing aya-ebpf-cty..."

Reply:
I think it's a good idea to avoid much discussion of portability, yes, because a lot of people will hear "portable bytecode VM" and assume it might be as portable as, say, wasm32.

wasm32, for instance, doesn't even have endianness problems. ^^;